truelearn/learning/_base.py

```diff
@@ -0,0 +1,565 @@
+        self.__calculate_draw_proba()
+        trueskill.setup(
+            mu=0.0,
+            sigma=InterestNoveltyKnowledgeBaseClassifier.DEFAULT_CONTENT_SIGMA,
+            beta=self.beta,
+            tau=self.tau,
+            draw_probability=self.draw_probability,
+            backend="mpmath",
+            env=self.__env,
+        )
+
+    @final
+    def __update_engagement_stats(self, y: bool) -> None:
+        """Update the learner's engagement stats based on the given label.
+
+        Args:
+            y: A bool indicating whether the learner engage in the learning event.
+        """
+        if y:
+            self.learner_model.number_of_engagements += 1
+        else:
+            self.learner_model.number_of_non_engagements += 1
+
+    @final
+    def _gather_trueskill_team(
+        self, kcs: Iterable[AbstractKnowledgeComponent]
+    ) -> Tuple[trueskill.Rating]:
+        """Return a tuple of trueskill Rating \
+        created from the given iterable of knowledge components.
+
+        Args:
+            kcs: An iterable of knowledge components.
+
+        Returns:
+            A tuple of trueskill Rating objects
+            created from the given iterable of knowledge components.
+        """
+        return tuple(
+            map(
+                lambda kc: self.__env.create_rating(
+                    mu=kc.mean, sigma=math.sqrt(kc.variance)
+                ),
+                kcs,
+            )
+        )
+
+    @abstractmethod
+    def _update_knowledge_representation(self, x: EventModel, y: bool) -> None:
+        """Update the knowledge representation of the LearnerModel.
+
+        Args:
+          x: A representation of the learning event.
+          y: A bool indicating whether the learner engages in the learning event.
+        """
+
+    @final
+    def fit(self, x: EventModel, y: bool) -> Self:
+        # if positive_only is False or (it's true and y is true)
+        # update the knowledge representation
+        if not self.positive_only or y is True:
+            self.__setup_env()
+            self._update_knowledge_representation(x, y)
+
+        self.__update_engagement_stats(y)
+        return self
+
+    @final
+    def predict(self, x: EventModel) -> bool:
+        return self.predict_proba(x) > self.threshold
+
+
+def team_sum_quality(
+    learner_kcs: Iterable[AbstractKnowledgeComponent],
+    content_kcs: Iterable[AbstractKnowledgeComponent],
+    beta: float,
+) -> float:
+    """Return the probability that the learner engages with the learnable unit.
+
+    Args:
+        learner_kcs: An iterable of knowledge components that come from the learner.
+        content_kcs: An iterable of knowledge components that come from the content.
+        beta: The noise factor.
+
+    Returns:
+        The probability that the learner engages with the learnable unit.
+    """
+    team_learner_mean = map(lambda kc: kc.mean, learner_kcs)
```

Comment:
Complains of duplication with these 6 lines in base(470-475) and lines 170-175 in knowledge classifier, I assume this is unavoidable and desired

Reply:
Perhaps you are referring to an outdated version of knowledge classifier?

Reply:
> Perhaps you are referring to an outdated version of knowledge classifier?

Nope Im all up to date:
![image](https://user-images.githubusercontent.com/91730569/218592774-ecd2ac30-b187-44f5-8613-58bd1feda91e.png)

Reply:
In the latest commit, there are only 161 lines in the knowledge classifier.

Reply:
> In the latest commit, there are only 161 lines in the knowledge classifier.

My bad I know why, I had an old build that I hadnt removed:
![image](https://user-images.githubusercontent.com/91730569/218596744-87bcddb9-553b-4c18-9290-b1da7ba32385.png)
 and it was finding the duplication that way

Reply:
😂 at least we solved this problem.